dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/permission/ManagePermissionsService.java

```diff
@@ -49,4 +98,39 @@ public DataverseRole saveOrUpdateRole(DataverseRole role) {
     public Dataverse setDataverseDefaultContributorRole(DataverseRole defaultRole, Dataverse affectedDataverse) {
         return commandEngine.submit(new UpdateDataverseDefaultContributorRoleCommand(defaultRole, dvRequestService.getDataverseRequest(), affectedDataverse));
     }
+
+    // -------------------- PRIVATE ---------------------
+    /**
+     * Notify a {@code RoleAssignee} that a role was either assigned or revoked.
+     * Will notify all members of a group.
+     *
+     * @param ra   The {@code RoleAssignee} to be notified.
+     * @param type The type of notification.
+     */
+    private void notifyRoleChange(RoleAssignee ra, DvObject dvObject, NotificationType type) {
+        if (ra instanceof AuthenticatedUser) {
+            userNotificationService.sendNotificationWithEmail((AuthenticatedUser) ra, new Timestamp(new Date().getTime()), type, dvObject.getId(), determineObjectType(dvObject));
+        } else if (ra instanceof ExplicitGroup) {
+            ExplicitGroup eg = (ExplicitGroup) ra;
+            Set<String> explicitGroupMembers = eg.getContainedRoleAssgineeIdentifiers();
```

Comment:
IMO this is more readable :P 
```
eg.getContainedRoleAssgineeIdentifiers().stream()
                    .map(roleId -> roleAssigneeService.getRoleAssignee(roleId))
                    .filter(roleAssignee -> roleAssignee instanceof AuthenticatedUser)
                    .forEach(roleAssignee -> userNotificationService.sendNotificationWithEmail());```

Reply:
I wouldn't try to make every bit of non-funcional-like code funcional-like :P

Reply:
Yeah not in every case, but usually for me it is more readable :P, in this case there is also a problem with class design.